File: test/Models/pheno_pkg/src/f90/Gaimean.f90
MODULE Gaimeanmod
    USE list_sub
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_gaimean(gAI, &
        tTWindowForPTQ, &
        deltaTT, &
        pastMaxAI_t1, &
        listTTShootWindowForPTQ1_t1, &
        listGAITTWindowForPTQ_t1, &
        gAImean, &
        pastMaxAI, &
        listTTShootWindowForPTQ1, &
        listGAITTWindowForPTQ)
        IMPLICIT NONE
        REAL, INTENT(IN) :: gAI
        REAL, INTENT(IN) :: tTWindowForPTQ
        REAL, INTENT(IN) :: deltaTT
        REAL, INTENT(IN) :: pastMaxAI_t1
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(IN) ::  &
                listTTShootWindowForPTQ1_t1
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(IN) ::  &
                listGAITTWindowForPTQ_t1
        REAL, INTENT(OUT) :: gAImean
        REAL, INTENT(OUT) :: pastMaxAI
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(OUT) ::  &
                listTTShootWindowForPTQ1
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(OUT) :: listGAITTWindowForPTQ
        REAL, ALLOCATABLE , DIMENSION(:):: TTList
        REAL, ALLOCATABLE , DIMENSION(:):: GAIList
        REAL:: SumTT
        INTEGER:: count
        REAL:: gai_
        REAL:: gaiMean_
        INTEGER:: countGaiMean
        INTEGER:: i
        count = 0
        gai_ = 0.0
        gaiMean_ = 0.0
        countGaiMean = 0
        !- Description:
    !            * Title: Average GAI on a specific thermal time window
    !            * Author: Loïc Manceau
    !            * Reference: -
    !            * Institution: INRA
    !            * Abstract: -
        !- inputs:
    !            * name: gAI
    !                          ** description : Green Area Index of the day
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 500.0
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
    !            * name: tTWindowForPTQ
    !                          ** description : Thermal Time window for average
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : °C d
    !                          ** uri : 
    !            * name: deltaTT
    !                          ** description : Thermal time increase of the day
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 100.0
    !                          ** unit : °C d
    !                          ** uri : 
    !            * name: pastMaxAI_t1
    !                          ** description : Maximum Leaf Area Index reached the current day
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
    !            * name: listTTShootWindowForPTQ1_t1
    !                          ** description : List of daily shoot thermal time in the window dedicated to average
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** default : [0.0]
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : °C d
    !                          ** uri : 
    !            * name: listGAITTWindowForPTQ_t1
    !                          ** description : List of daily Green Area Index in the window dedicated to average
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** default : [0.0]
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
        !- outputs:
    !            * name: gAImean
    !                          ** description : Mean Green Area Index
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 500.0
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
    !            * name: pastMaxAI
    !                          ** description : Maximum Leaf Area Index reached the current day
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
    !            * name: listTTShootWindowForPTQ1
    !                          ** description : List of daily shoot thermal time in the window dedicated to average
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : °C d
    !                          ** uri : 
    !            * name: listGAITTWindowForPTQ
    !                          ** description : List of daily Green Area Index in the window dedicated to average
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : m2 leaf m-2 ground
    !                          ** uri : 
        DO i = 0 , SIZE(listTTShootWindowForPTQ1_t1)-1, 1
            call Add(TTList, listTTShootWindowForPTQ1_t1(i+1))
            call Add(GAIList, listGAITTWindowForPTQ_t1(i+1))
        END DO
        call Add(TTList, deltaTT)
        call Add(GAIList, gAI)
        SumTT = sum(TTList)
        DO WHILE ( SumTT .GT. tTWindowForPTQ )
            SumTT = SumTT - TTList(count+1)
            count = count + 1
        END DO
        DO i = count , SIZE(TTList)-1, 1
            call Add(listTTShootWindowForPTQ1, TTList(i+1))
            call Add(listGAITTWindowForPTQ, GAIList(i+1))
        END DO
        DO i = 0 , SIZE(listGAITTWindowForPTQ)-1, 1
            gaiMean_ = gaiMean_ + listGAITTWindowForPTQ(i+1)
            countGaiMean = countGaiMean + 1
        END DO
        gaiMean_ = gaiMean_ / countGaiMean
        gai_ = MAX(pastMaxAI_t1, gaiMean_)
        pastMaxAI = gai_
        gAImean = gai_
    END SUBROUTINE model_gaimean

END MODULE

File: test/Models/pheno_pkg/src/openalea/Gaimean.py
# coding: utf8
from copy import copy

import numpy
from math import *

def model_gaimean(gAI = 0.0,
         tTWindowForPTQ = 0.0,
         deltaTT = 0.0,
         pastMaxAI_t1 = 0.0,
         listTTShootWindowForPTQ1_t1 = [0.0],
         listGAITTWindowForPTQ_t1 = [0.0]):
    """
     - Name: GAImean -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Average GAI on a specific thermal time window
                 * Author: Loïc Manceau
                 * Reference: -
                 * Institution: INRA
                 * Abstract: -
     - inputs:
                 * name: gAI
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 500.0
                               ** uri : 
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : m2 leaf m-2 ground
                               ** description : Green Area Index of the day
                 * name: tTWindowForPTQ
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 5000.0
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : °C d
                               ** description : Thermal Time window for average
                 * name: deltaTT
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : 
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : Thermal time increase of the day
                 * name: pastMaxAI_t1
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 5000.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : m2 leaf m-2 ground
                               ** description : Maximum Leaf Area Index reached the current day
                 * name: listTTShootWindowForPTQ1_t1
                               ** min : 
                               ** default : [0.0]
                               ** max : 
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : List of daily shoot thermal time in the window dedicated to average
                 * name: listGAITTWindowForPTQ_t1
                               ** min : 
                               ** default : [0.0]
                               ** max : 
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** inputtype : variable
                               ** unit : m2 leaf m-2 ground
                               ** description : List of daily Green Area Index in the window dedicated to average
     - outputs:
                 * name: gAImean
                               ** min : 0.0
                               ** variablecategory : state
                               ** max : 500.0
                               ** uri : 
                               ** datatype : DOUBLE
                               ** unit : m2 leaf m-2 ground
                               ** description : Mean Green Area Index
                 * name: pastMaxAI
                               ** min : 0.0
                               ** variablecategory : state
                               ** max : 5000.0
                               ** uri : 
                               ** datatype : DOUBLE
                               ** unit : m2 leaf m-2 ground
                               ** description : Maximum Leaf Area Index reached the current day
                 * name: listTTShootWindowForPTQ1
                               ** min : 
                               ** variablecategory : state
                               ** max : 
                               ** uri : 
                               ** datatype : DOUBLELIST
                               ** unit : °C d
                               ** description : List of daily shoot thermal time in the window dedicated to average
                 * name: listGAITTWindowForPTQ
                               ** min : 
                               ** variablecategory : state
                               ** max : 
                               ** uri : 
                               ** datatype : DOUBLELIST
                               ** unit : m2 leaf m-2 ground
                               ** description : List of daily Green Area Index in the window dedicated to average
    """

    listTTShootWindowForPTQ1 = []
    listGAITTWindowForPTQ = []
    TTList = []
    GAIList = []
    count = 0
    gai_ = 0.0
    gaiMean_ = 0.0
    countGaiMean = 0
    for i in range(0 , len(listTTShootWindowForPTQ1_t1) , 1):
        TTList.append(listTTShootWindowForPTQ1_t1[i])
        GAIList.append(listGAITTWindowForPTQ_t1[i])
    TTList.append(deltaTT)
    GAIList.append(gAI)
    SumTT = sum(TTList)
    while SumTT > tTWindowForPTQ:
        SumTT = SumTT - TTList[count]
        count = count + 1
    for i in range(count , len(TTList) , 1):
        listTTShootWindowForPTQ1.append(TTList[i])
        listGAITTWindowForPTQ.append(GAIList[i])
    for i in range(0 , len(listGAITTWindowForPTQ) , 1):
        gaiMean_ = gaiMean_ + listGAITTWindowForPTQ[i]
        countGaiMean = countGaiMean + 1
    gaiMean_ = gaiMean_ / countGaiMean
    gai_ = max(pastMaxAI_t1, gaiMean_)
    pastMaxAI = gai_
    gAImean = gai_
    return (gAImean, pastMaxAI, listTTShootWindowForPTQ1, listGAITTWindowForPTQ)